### PR TITLE
MM-23703 Use thread for User Satisfaction Surveybot messages

### PR DESCRIPTION
**Pull request by Lumexralph**
#### Summary

This PR helps to ensure the bot responds to the user in a thread as opposed to the initial behaviour where every response is sent to the channel.
#### Ticket Link

Fixes [14255](https://github.com/mattermost/mattermost-server/issues/14255)

### CODE REVIEW

File: server/hooks.go
Comment (hanzei) [inline server/hooks.go:69]: As far as I understand the ticket, this is not the correct logic. I think it should be
- If the user did reply to the existing post, also rely to this post in a thread.
- If the user didn't rely to the existing post i.e. posted a new message in the main panel, post the response also in the main panel.

@esadur Can you confirm that my understanding of the ticket is correct?

Comment (esethna) [reply server/hooks.go:69]: @hanzei not quite, based on the ticket spec: "the responses as a reply to the users feedback". So this means

1. If the user did reply to the existing post, also rely to this post in the thread.
2. If the user didn't rely to the existing post i.e. posted a new message in the main panel, post the surveybot response **as a reply to the users message**

**Comment by esethna** [comment]
@hmhealey @hanzei how can we test this?

**Comment by hmhealey** [comment]
@esethna I can prepare a custom build of the plugin if we want, but this logic is simple enough that it can be covered by unit tests.

**Comment by esethna** [comment]
I'll leave that up to QA ^ cc// @jgilliam17

**Comment by ogi-m** [comment]
We're fine to merge this with unit test coverage and QA can test on the qa branch after that.

**Comment by Lumexralph** [comment]
> Thanks for the PR, @Lumexralph. The code for this looks good, but would you be able to add some unit tests for this behaviour?
> 
> You'll want to make changes to [`TestMessageHasBeenPosted`](https://github.com/mattermost/mattermost-plugin-nps/blob/master/server/hooks_test.go#L42) to do that. That function is quite complicated because of the logic we have to mock out the server, so I could help give advice on how to add those tests if you'd like.

Thanks @hmhealey I'll reach out to you on how to go about the unit tests.

**Comment by esethna** [comment]
@Lumexralph can you help confirm the following behaviour (cc// @ogi-m) :

1. If the user did reply to the existing post, also rely to this post in the thread.
2. If the user didn't rely to the existing post i.e. posted a new message in the main panel, post the surveybot response as a reply to the users message

If so, @hanzei feel free to remove the PM review label (I don't have the permissions on this repo)

**Comment by hmhealey** [comment]
Hey @Lumexralph. How's it going? Is there anything we can help with on this?

**Comment by Lumexralph** [comment]
> Hey @Lumexralph. How's it going? Is there anything we can help with on this?

Not so much, I just need to update the test with the already existing mocks, and also there was a previous discussion about checking that the new change addressed the issue to be fixed. This actually made me wait to see if anyone tried it out.

Should I proceed to add the test?

**Comment by jasonblais** [comment]
cc @hmhealey on the above question regarding tests

File: server/hooks.go
Comment (hmhealey) [reply server/hooks.go:69]: This logic is correct then. If the user posted in a thread (`post.RootId` is not empty), then reply to the same thread (set `rootID` to `post.RootId`), otherwise, reply to the user's post (set `rootID` to `post.Id`).

**Comment by hmhealey** [comment]
I resolved that conversation since the existing logic looks correct to me. Sorry for any confusion around that.

If you could please add the test, that'd be great. The complicated mocking I was referring to is setting up all the `api.On` calls correctly like in the other tests. You'll need to modify the mock for `CreatePost` to ensure the provided post's `RootId` is set correctly, and I think that can be done by using [`mock.MatchedBy`](https://gowalker.org/github.com/stretchr/testify/mock#MatchedBy) like [this](https://github.com/mattermost/mattermost-plugin-nps/blob/d907b2bad27460bb779392ac6a4e62d7b9f2b06a/server/hooks_test.go#L18)

**Comment by jasonblais** [comment]
Hey @Lumexralph, let us know if you have any questions about adding the test, we'd be happy to help :)

**Comment by Lumexralph** [comment]
> Hey @Lumexralph, let us know if you have any questions about adding the test, we'd be happy to help :)

Hi @jasonblais , thanks for the patience, I'll get the test in by the end of the week. I've got a busy week here. I hope that's fine?

**Comment by jasonblais** [comment]
@Lumexralph Certainly! Thank you for the heads up :)

**Review by hanzei — APPROVED** [review]
LGMT :+1: Nice work on the test

**Comment by hanzei** [comment]
@Lumexralph Heads up that there is a merge conflict.

**Comment by Lumexralph** [comment]
> @Lumexralph Heads up that there is a merge conflict.

@hanzei I will resolve the conflicts maybe later today. Thanks for your patience.

**Comment by Lumexralph** [comment]
> @Lumexralph Certainly! Thank you for the heads up :)

Hi again, I tried updating my branch with the new change, I keep running into this issue when I run the test.

```
panic: 

mock: Unexpected Method Call
-----------------------------

GetTeamMembersForUser(string,int,int)
		0: ""
		1: 0
		2: 50

The closest call I have is: 

GetTeamMembersForUser(string,int,int)
		0: "icixz5p8ufnkjd3pnwgxdrz3mo"
		1: 0
		2: 50


Diff: 0: FAIL:  (string=) != (string=icixz5p8ufnkjd3pnwgxdrz3mo)
	1: PASS:  (int=0) == (int=0)
	2: PASS:  (int=50) == (int=50) [recovered]
	panic: 

mock: Unexpected Method Call
-----------------------------

GetTeamMembersForUser(string,int,int)
		0: ""
		1: 0
		2: 50

The closest call I have is: 

GetTeamMembersForUser(string,int,int)
		0: "icixz5p8ufnkjd3pnwgxdrz3mo"
		1: 0
		2: 50


Diff: 0: FAIL:  (string=) != (string=icixz5p8ufnkjd3pnwgxdrz3mo)
	1: PASS:  (int=0) == (int=0)
	2: PASS:  (int=50) == (int=50)

goroutine 8 [running]:
testing.tRunner.func1(0xc000358500)
	/usr/local/Cellar/go/1.13.5/libexec/src/testing/testing.go:874 +0x3a3
panic(0x18a7a00, 0xc0004040f0)
	/usr/local/Cellar/go/1.13.5/libexec/src/runtime/panic.go:679 +0x1b2
github.com/stretchr/testify/mock.(*Mock).fail(0xc0000328c0, 0x1a60cd1, 0x6f, 0xc000095ec0, 0x4, 0x4)
	/Users/olumideogundele/go/pkg/mod/github.com/stretchr/testify@v1.6.1/mock/mock.go:254 +0x1a0
github.com/stretchr/testify/mock.(*Mock).MethodCalled(0xc0000328c0, 0x211bab4, 0x15, 0xc0003b9dd0, 0x3, 0x3, 0xc000095d40, 0x4, 0x4)
	/Users/olumideogundele/go/pkg/mod/github.com/stretchr/testify@v1.6.1/mock/mock.go:384 +0x65c
github.com/stretchr/testify/mock.(*Mock).Called(0xc0000328c0, 0xc0003b9dd0, 0x3, 0x3, 0x7, 0xc000081de0, 0x1)
	/Users/olumideogundele/go/pkg/mod/github.com/stretchr/testify@v1.6.1/mock/mock.go:356 +0x1a7
github.com/mattermost/mattermost-server/v5/plugin/plugintest.(*API).GetTeamMembersForUser(0xc0000328c0, 0x0, 0x0, 0x0, 0x32, 0xc0001218c8, 0x17d78e2, 0x0, 0x0)
	/Users/olumideogundele/go/pkg/mod/github.com/mattermost/mattermost-server/v5@v5.3.2-0.20200710040900-da95d40e3338/plugin/plugintest/api.go:1736 +0x127
github.com/mattermost/mattermost-plugin-nps/server/server.(*Plugin).isUserTeamAdmin(0xc000121d88, 0xc000001980, 0xc0003cf320)
	/Users/olumideogundele/go/src/github.com/mattermost/mattermost-plugin-nps/server/telemetry.go:95 +0x10e
github.com/mattermost/mattermost-plugin-nps/server/server.(*Plugin).getUserRole(0xc000121d88, 0xc000001980, 0x1a, 0xc000001980)
	/Users/olumideogundele/go/src/github.com/mattermost/mattermost-plugin-nps/server/telemetry.go:83 +0x4e
github.com/mattermost/mattermost-plugin-nps/server/server.(*Plugin).getEventProperties(0xc000121d88, 0xc0003daf20, 0x1a, 0x0, 0xc000121a00, 0x0)
	/Users/olumideogundele/go/src/github.com/mattermost/mattermost-plugin-nps/server/telemetry.go:61 +0x56a
github.com/mattermost/mattermost-plugin-nps/server/server.(*Plugin).sendFeedback(0xc000121d88, 0x0, 0x0, 0xc0003daf20, 0x1a, 0x0)
	/Users/olumideogundele/go/src/github.com/mattermost/mattermost-plugin-nps/server/telemetry.go:41 +0x134
github.com/mattermost/mattermost-plugin-nps/server/server.(*Plugin).MessageHasBeenPosted(0xc000121d88, 0x0, 0xc000121e20)
	/Users/olumideogundele/go/src/github.com/mattermost/mattermost-plugin-nps/server/hooks.go:59 +0x1c1
github.com/mattermost/mattermost-plugin-nps/server/server.TestMessageHasBeenPosted.func1(0xc000358500)
	/Users/olumideogundele/go/src/github.com/mattermost/mattermost-plugin-nps/server/hooks_test.go:92 +0xdaf
testing.tRunner(0xc000358500, 0xc0003d1290)
	/usr/local/Cellar/go/1.13.5/libexec/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/local/Cellar/go/1.13.5/libexec/src/testing/testing.go:960 +0x350
FAIL	github.com/mattermost/mattermost-plugin-nps/server/server	0.916s
FAIL
```

The cause is because of these 2 subtests, if I remove the second subtest, the test passes:

```
	t.Run("should send feedback to segment and respond to user's post on existing post", func(t *testing.T) {
		api := &plugintest.API{}
		api.On("GetConfig").Return(&model.Config{
			LogSettings: model.LogSettings{
				EnableDiagnostics: model.NewBool(true),
			},
		})
		api.On("GetChannel", botChannelID).Return(&model.Channel{
			Type: model.CHANNEL_DIRECT,
			Name: fmt.Sprintf("%s__%s", botUserID, userID),
		}, nil)
		api.On("GetUser", userID).Return(&model.User{Id: userID}, nil)
		api.On("GetDirectChannel", userID, botUserID).Return(&model.Channel{
			Id: botChannelID,
		}, nil)
		api.On("CreatePost", mock.MatchedBy(func(post *model.Post) bool {
			return post.RootId == rootID
		})).Return(nil, nil)
		api.On("GetSystemInstallDate").Return(systemInstallDate, nil)
		api.On("GetTeamMembersForUser", userID, 0, 50).Return(teamMembers, nil)
		api.On("GetLicense").Return(&model.License{
			Id:           licenseID,
			SkuShortName: skuShortName,
		})
		defer api.AssertExpectations(t)

		p := &Plugin{
			botUserID: botUserID,
			tracker:   telemetry.NewTracker(nil, "", "", "", "", "", false, nil),
		}
		p.SetAPI(api)

		p.MessageHasBeenPosted(nil, &model.Post{
			ChannelId: botChannelID,
			UserId:    userID,
			RootId:    rootID,
		})
	})

	postID := model.NewId()

	t.Run("should send feedback to segment and respond to user's new post", func(t *testing.T) {
		api := &plugintest.API{}
		api.On("GetConfig").Return(&model.Config{
			LogSettings: model.LogSettings{
				EnableDiagnostics: model.NewBool(true),
			},
		})
		api.On("GetChannel", botChannelID).Return(&model.Channel{
			Type: model.CHANNEL_DIRECT,
			Name: fmt.Sprintf("%s__%s", botUserID, userID),
		}, nil)
		api.On("GetUser", userID).Return(&model.User{}, nil)
		api.On("GetDirectChannel", userID, botUserID).Return(&model.Channel{
			Id: botChannelID,
		}, nil)
		api.On("CreatePost", mock.MatchedBy(func(post *model.Post) bool {
			return post.RootId == postID
		})).Return(nil, nil)
		defer api.AssertExpectations(t)

		p := &Plugin{
			botUserID: botUserID,
			tracker:   telemetry.NewTracker(nil, "", "", "", "", "", false, nil),
		}
		p.SetAPI(api)

		p.MessageHasBeenPosted(nil, &model.Post{
			Id:        postID,
			ChannelId: botChannelID,
			UserId:    userID,
		})
	})
```

How do you think I can resolve this?

**Comment by jasonblais** [comment]
^cc @hanzei on above, if you're open to helping when you have a moment?

**Comment by hmhealey** [comment]
Those "Unexpected Method Call" errors come from the mocking code that's pretending to be the server's plugin API when you either haven't set up a mock for a given method or you've given it incorrect arguments. For example, in the first one

> ```
> panic: 
> 
> mock: Unexpected Method Call
> -----------------------------
> 
> GetTeamMembersForUser(string,int,int)
> 		0: ""
> 		1: 0
> 		2: 50
> 
> The closest call I have is: 
> 
> GetTeamMembersForUser(string,int,int)
> 		0: "icixz5p8ufnkjd3pnwgxdrz3mo"
> 		1: 0
> 		2: 50
> ```

You told it to expect a call to `GetTeamMembersForUser` where the first argument is a user ID, but it was called with the first argument as the empty string, so either the code isn't passing a user ID correctly or the mock shouldn't be expecting an ID there.

It looks like the tests on master are passing for me, so perhaps you've merged something incorrectly?

**Comment by Lumexralph** [comment]
> Those "Unexpected Method Call" errors come from the mocking code that's pretending to be the server's plugin API when you either haven't set up a mock for a given method or you've given it incorrect arguments. For example, in the first one
> 
> > ```
> > panic: 
> > 
> > mock: Unexpected Method Call
> > -----------------------------
> > 
> > GetTeamMembersForUser(string,int,int)
> > 		0: ""
> > 		1: 0
> > 		2: 50
> > 
> > The closest call I have is: 
> > 
> > GetTeamMembersForUser(string,int,int)
> > 		0: "icixz5p8ufnkjd3pnwgxdrz3mo"
> > 		1: 0
> > 		2: 50
> > ```
> 
> You told it to expect a call to `GetTeamMembersForUser` where the first argument is a user ID, but it was called with the first argument as the empty string, so either the code isn't passing a user ID correctly or the mock shouldn't be expecting an ID there.
> 
> It looks like the tests on master are passing for me, so perhaps you've merged something incorrectly?

Everything worked fine but I had merge conflicts to resolve because new changes were added and that affected the test and like I said earlier, if I remove the second subtests, everything works fine and the test is also needed.

I just want to know how I can fix this, it is not a code I wrote, it was as a result of the code that was added to `hooks.go`, `hooks_test.go` during `merge`

**Comment by hanzei** [comment]
@Lumexralph Could you please merge the changes and push, even if that breakes CI? That would make it easier to debug the failure.

**Comment by Lumexralph** [comment]
> @Lumexralph Could you please merge the changes and push, even if that breakes CI? That would make it easier to debug the failure.

@hanzei I just pushed the new changes, thanks.

**Comment by hanzei** [comment]
@Lumexralph Thanks for pushing your changes! I've fixed the test myself. Sorry for the incontinence.

**Comment by esethna** [comment]
Let's include this PR in the NPS v1.1 push: https://mattermost.atlassian.net/browse/MM-28520

**Comment by esethna** [comment]
@hmhealey are we good to merge this? cc// @ogi-m on testing after the merge (I think that's the process)

**Comment by hmhealey** [comment]
Yeah, if we're good with testing after merge, that should be fine.

**Pull request merged**
